File: code_samples/section7/problems/problem7_5_1/problem7_5_1.java
package code_samples.section7.problems.problem7_5_1;

import java.util.*;

/**
 * problem7_5_1
 *
 * Demonstrates a hash-map-based solution to the classic
 * "Two Sum (return indices)" problem.
 *
 * Given an array of integers and a target value, the goal is to find
 * two distinct indices i and j such that:
 *
 *     nums[i] + nums[j] == target
 *
 * If no such pair exists, an empty array is returned.
 */
public class problem7_5_1 {

    /**
     * twoSumIndices
     *
     * Finds indices of two elements in the array whose values sum to `target`.
     *
     * @param nums   Input array of integers.
     * @param target Target sum.
     *
     * @return An array of two indices [i, j] such that
     *         nums[i] + nums[j] == target,
     *         or an empty array if no solution exists.
     *
     * Approach (single-pass hash map):
     *   - Use a HashMap to store values we have already seen along with
     *     their indices.
     *   - As we iterate through the array:
     *       1) Compute the value needed to reach the target.
     *       2) If that value is already in the map, we have found a valid pair.
     *       3) Otherwise, store the current value and its index in the map.
     *
     * Why this works:
     *   - At index i, the map contains only elements from indices < i,
     *     guaranteeing the two indices are distinct.
     *
     * Complexity (average / expected):
     *   - Time:  O(n)
     *       Each element is processed once, with O(1) average-time map operations.
     *   - Space: O(n)
     *       The map may store up to n elements.
     *
     * Notes:
     *   - If the same value appears multiple times, the map keeps the
     *     most recently stored index for that value.
     *   - This implementation returns the first valid pair it encounters.
     */
    int[] twoSumIndices(int[] nums, int target) {
        // Hash map: value -> index where that value appears
        Map<Integer, Integer> mp = new HashMap<>();

        // Iterate through the array once
        for (int i = 0; i < nums.length; i++) {
            // Compute the complement needed to reach the target
            int need = target - nums[i];

            // If we've already seen the needed value, we have a solution
            if (mp.containsKey(need)) {
                // Return indices of the matching pair
                return new int[]{mp.get(need), i};
            }

            // Otherwise, record the current value and its index
            mp.put(nums[i], i);
        }

        // No valid pair found
        return new int[0];
    }

    // ===============================
    // Test Harness
    // ===============================

    /**
     * main
     *
     * Simple console-based test harness for twoSumIndices().
     * Runs several test cases and prints both actual and expected results.
     */
    public static void main(String[] args) {
        problem7_5_1 sol = new problem7_5_1();

        // Test input arrays
        int[][] tests = {
            {2, 7, 11, 15},   // expected [0,1]
            {3, 2, 4},        // expected [1,2]
            {1, 2, 3, 4}      // expected [] (no solution)
        };

        // Corresponding target values
        int[] targets = {9, 6, 10};

        // Expected outputs for each test
        int[][] expected = {
            {0, 1},
            {1, 2},
            {}                 // no result
        };

        System.out.println("=== Test: twoSumIndices ===\n");

        // Run each test case
        for (int t = 0; t < tests.length; t++) {
            int[] nums = tests[t];
            int target = targets[t];

            // Call the function under test
            int[] result = sol.twoSumIndices(nums, target);

            // Print results
            System.out.println(
                "Input nums = " + Arrays.toString(nums) +
                ", target = " + target
            );
            System.out.println(
                "Result: " + Arrays.toString(result) +
                " (expected " + Arrays.toString(expected[t]) + ")\n"
            );
        }
    }
}

File: code_samples/section7/problems/problem7_5_1/problem7_5_1.cs
/**
 * TwoSumIndices
 *
 * Purpose:
 *   Finds two distinct indices i and j in the array `nums` such that:
 *       nums[i] + nums[j] == target
 *
 * Parameters:
 *   nums   - Input array of integers.
 *   target - Target sum.
 *
 * Returns:
 *   An array of two indices [i, j] (here returned as [j, i] based on lookup order),
 *   or an empty array [] if no solution exists.
 *
 * Approach (single-pass hash table):
 *   - Use a Dictionary<int, int> to map:
 *       value -> index where that value was last seen
 *   - Iterate through nums:
 *       1) Compute need = target - nums[i]
 *       2) If need is already in the map, we've found a prior value that pairs
 *          with nums[i], so return its index and i.
 *       3) Otherwise, record nums[i] -> i and continue.
 *
 * Why this works:
 *   - Any time we are at position i, all indices < i have already been recorded
 *     in the dictionary, so finding `need` guarantees we found a distinct earlier
 *     element.
 *
 * Complexity (expected / average):
 *   - Time:  O(n)
 *       Each element does O(1) average dictionary operations.
 *   - Space: O(n)
 *       The dictionary may store up to n entries.
 *
 * Notes:
 *   - If the same value appears multiple times, the dictionary stores the most
 *     recent index for that value (because we overwrite map[nums[i]] = i).
 *     This is fine for the classic "return any one valid pair" Two Sum problem.
 *   - The array literal syntax [j, i] and [] uses collection expressions (C# 12+).
 */
static int[] TwoSumIndices(int[] nums, int target) {
    // Map: value -> index
    var map = new Dictionary<int, int>();

    // Iterate through the array once
    for (int i = 0; i < nums.Length; i++) {
        // Complement we need to reach target using nums[i]
        int need = target - nums[i];

        // If we've already seen `need`, we have a solution
        if (map.TryGetValue(need, out int j)) {
            return [j, i];
        }

        // Record (or update) the index for this value
        map[nums[i]] = i;
    }

    // No solution found
    return [];
}

// ============================
// Top-level tests
// ============================
//
// These are top-level statements (C# 9+). The array literals used here
// are collection expressions (C# 12+).
//

int[][] tests =
[
    [2, 7, 11, 15],  // expected [0,1]
    [3, 2, 4],       // expected [1,2]
    [1, 2, 3, 4]     // expected []
];

int[] targets = [9, 6, 10];

int[][] expected =
[
    [0, 1],
    [1, 2],
    []
];

Console.WriteLine("=== Test: TwoSumIndices ===\n");

// Run each test case
for (int t = 0; t < tests.Length; t++)
{
    int[] nums = tests[t];
    int target = targets[t];

    // Call the function under test
    int[] result = TwoSumIndices(nums, target);

    // Pretty-print results
    string resultStr = "[" + string.Join(", ", result) + "]";
    string expectedStr = "[" + string.Join(", ", expected[t]) + "]";

    Console.WriteLine($"Input nums = [{string.Join(", ", nums)}], target = {target}");
    Console.WriteLine($"Result: {resultStr} (expected {expectedStr})\n");
}
